### PR TITLE
Implement GlrawTextureLoader

### PR DESCRIPTION
**Pull request by chistopher**
GlrawLoader as requested in issue #13.
Most code is from [cginternals/glraw](https://github.com/cginternals/glraw). 
Needs [cppassist#35](https://github.com/cginternals/cppassist/pull/35)

- [x] load .raw
- [x] load .glraw
- [x] remove qt usage

### CODE REVIEW

File: source/plugins/gloperate-glraw-loaders/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+# 
+
+find_package(cpplocate REQUIRED)
+find_package(cppassist REQUIRED)
+find_package(cppexpose REQUIRED)
+find_package(glbinding REQUIRED)
+find_package(globjects REQUIRED)
+
+# Qt5
+find_package(Qt5Core 5.1)
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/CMakeLists.txt]: As Qt is not a dependency anymore, these lines can be removed (up to line 25).

File: source/plugins/gloperate-glraw-loaders/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+# 
+# Library name and options
+# 
+
+# Target name
+set(target gloperate-glraw-loaders)
+
+# Exit here if required dependencies are not met
+if (NOT Qt5Core_FOUND)
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/CMakeLists.txt]: Update required here.

File: source/plugins/gloperate-glraw-loaders/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    PRIVATE
+    Qt5::Core
+    Qt5::Gui
+    cpplocate::cpplocate
+    cppassist::cppassist
+    cppexpose::cppexpose
+    glbinding::glbinding
+    globjects::globjects
+    ${META_PROJECT_NAME}::gloperate
+    ${META_PROJECT_NAME}::gloperate-qt
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/CMakeLists.txt]: And here.

File: source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	{
+        { "ub", gl::GL_UNSIGNED_BYTE    },
+        { "b",  gl::GL_BYTE             },
+        { "us", gl::GL_UNSIGNED_SHORT   },
+        { "s",  gl::GL_SHORT            },
+        { "ui", gl::GL_UNSIGNED_INT     },
+        { "i",  gl::GL_INT              },
+        { "f",  gl::GL_FLOAT            }
+
+    #ifdef gl::GL_ARB_texture_compression_rgtc
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp]: This doesn't work as expected. Remove the ```#idef``` conditional code and add the formats in each case.

File: source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+
+namespace glraw
+{
+
+FileNameSuffix::FileNameSuffix(const std::string & fileName)
+: m_width (-1)
+, m_height(-1)
+, m_format(gl::GL_INVALID_ENUM)
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp]: ```GL_INVALID_ENUM``` is an error code and not an invalid enum value. Use ```GL_NONE``` (0x0) instead.

File: source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    }
+    catch(...)
+    {
+        return;
+    }
+
+	if (!m_compressed)
+	{
+        m_format = format(base_match[3]);
+        assert(m_format != gl::GL_INVALID_ENUM);
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp]: Test against updated value (e.g., ```GL_NONE```)

File: source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    m_type = type(base_match[m_compressed ? 3 : 4]);
+    assert(m_type != gl::GL_INVALID_ENUM);
+}
+
+
+bool FileNameSuffix::isValid() const
+{
+	return m_width  != -1
+		&& m_height != -1
+        && m_type   != gl::GL_INVALID_ENUM
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp]: Here, too.

File: source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+{
+	return m_suffix;
+}
+
+gl::GLenum FileNameSuffix::format(const std::string & format)
+{
+    std::string formatToLower = format;
+    std::transform(formatToLower.begin(), formatToLower.end(), formatToLower.begin(), ::tolower);
+    auto it = formatsBySuffix.find(formatToLower);
+    return (it!=formatsBySuffix.end()) ? it->second : gl::GL_INVALID_ENUM;
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp]: Here, too.

File: source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+
+
+gl::GLenum FileNameSuffix::type(const std::string & type)
+{
+    std::string typeToLower = type;
+    std::transform(typeToLower.begin(), typeToLower.end(), typeToLower.begin(), ::tolower);
+    auto it = typesBySuffix.find(typeToLower);
+    return (it!=typesBySuffix.end()) ? it->second : gl::GL_INVALID_ENUM;
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp]: And here :)

File: source/plugins/gloperate-glraw-loaders/FileNameSuffix.h [hunk 10]
```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+
+#include <glbinding/gl/gl.h>
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.h]: It should suffice to include ```glbinding/gl/types.h```.

File: source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp [hunk 11]
```diff
@@ -0,0 +1,159 @@
+#include <FileNameSuffix.h>
+
+#include <cassert>
+#include <regex>
+#include <string>
+
+#include <glbinding/gl/gl.h>
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp]: It should suffice to include ```glbinding/gl/enum.h```.

File: source/plugins/gloperate-glraw-loaders/GlrawTextureLoader.cpp [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+GlrawTextureLoader::~GlrawTextureLoader()
+{
+}
+
+bool GlrawTextureLoader::canLoad(const std::string & ext) const
+{
+    // Check if file type is supported
+    return (std::count(m_extensions.begin(), m_extensions.end(), "." + ext) > 0);
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/GlrawTextureLoader.cpp]: A ```std::find``` and a test against the end iterator should be quicker in most cases.

File: source/plugins/gloperate-glraw-loaders/GlrawTextureLoader.cpp [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+std::vector<std::string> GlrawTextureLoader::loadingTypes() const
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/GlrawTextureLoader.cpp]: Return const-references.

File: source/plugins/gloperate-glraw-loaders/GlrawTextureLoader.cpp [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+globjects::Texture * GlrawTextureLoader::loadRawImage(const std::string & filename) const
+{
+    glraw::FileNameSuffix suffix(filename);
+    if (!suffix.isValid())
+        return nullptr;
+
+    //read file
+    std::ifstream ifs(filename, std::ios::binary | std::ios::ate);
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/GlrawTextureLoader.cpp]: Why do you not use ```RawFile``` here?

File: source/plugins/gloperate-glraw-loaders/RawFile.cpp [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+            return;
+        }
+    }
+}
+
+void RawFile::readRawData(std::ifstream & ifs, uint64_t rawDataOffset)
+{
+    ifs.seekg(0, std::ios::end);
+    
+    size_t endPosition = ifs.tellg();
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/RawFile.cpp]: ```endPosition``` can be declared const, too.

File: source/plugins/gloperate-glraw-loaders/RawFile.h [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace glraw
+{
+
+class RawFile
+{
+public:
+    static uint16_t s_signature;
+
+    enum class PropertyType
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/RawFile.h]: Provide explicit size.

**Comment by scheibel** [comment]
Please adhere to our additional [source code style guide](http://cginternals.github.io/guidelines/articles/style-guide/) and [comment style guide](http://cginternals.github.io/guidelines/articles/comment-style-guide/).

File: source/plugins/gloperate-glraw-loaders/FileNameSuffix.h (hunk 10 above)
Comment (chistopher) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.h]: No. But `glbinding/gl/enum.h` does.

File: source/plugins/gloperate-glraw-loaders/GlrawTextureLoader.cpp (hunk 13 above)
Comment (chistopher) [inline source/plugins/gloperate-glraw-loaders/GlrawTextureLoader.cpp]: Sure? It's an override.

File: source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+, m_format(gl::GL_NONE)
+, m_type(gl::GL_NONE)
+, m_compressed(false)
+{
+	// check if either compressed or uncompressed (or unknown) format
+    std::regex regexp{R"(^.*\.(\d+)\.(\d+)\.(\w+)\.raw$)"};
+    std::smatch base_match;
+
+    if (std::regex_match(fileName, base_match, regexp))
+		m_compressed = true;
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp]: Please replace all indentation using TAB (\t) with 4 spaces.

File: source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+
+FileNameSuffix::FileNameSuffix(const std::string & fileName)
+: m_width (-1)
+, m_height(-1)
+, m_format(gl::GL_NONE)
+, m_type(gl::GL_NONE)
+, m_compressed(false)
+{
+	// check if either compressed or uncompressed (or unknown) format
+    std::regex regexp{R"(^.*\.(\d+)\.(\d+)\.(\w+)\.raw$)"};
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp]: The regex (and I assume all others) can be created statically. Further, consider moving the declarations to the anonymous namespace above.

File: source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+
+bool FileNameSuffix::compressed() const
+{
+    return m_compressed;
+}
+
+gl::GLenum FileNameSuffix::format(const std::string & format)
+{
+    std::string formatToLower = format;
+    std::transform(formatToLower.begin(), formatToLower.end(), formatToLower.begin(), ::tolower);
```
Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp]: I don't like copying strings for the case-insensitive lookup in a map. However, this seems to be a faster implementation than the alternative: a case-insensitive comparator on a map, where each compare would repeatedly perform the case conversion.

Comment (scheibel) [inline source/plugins/gloperate-glraw-loaders/FileNameSuffix.cpp]: But you may add a ```toLower``` and a ```toUpper``` function to cppassist ```string/conversion.h``` and use this function instead.

**Comment by scheibel** [comment]
One further task (and hopefully the last): please move the classes from the plugin to the gloperate main module.
Class destination:
* ```gloperate/base/FileNameSuffix```
* ```gloperate/loaders/GlrawLoader```

File: source/gloperate/source/base/RawFileNameSuffix.cpp [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+    else
+    {
+        if (!std::regex_match(fileName, base_match, regex2))
+            return;
+    }
+
+    // retrieve intel from suffix parts
+    try
+    {
+        m_width = std::stoi(base_match[1]);
```
Comment (scheibel) [inline source/gloperate/source/base/RawFileNameSuffix.cpp]: Use ```cppassist::string::fromString<int>``` here.

Comment (scheibel) [inline source/gloperate/source/base/RawFileNameSuffix.cpp]: Then, the `try...catch` block can be removed.

File: source/gloperate/include/gloperate/loaders/GlrawTextureLoader.h [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+    *    This function is called by load when the file extension is .raw
+    *
+    *  @param[in] filename
+    *    path of the .raw file
+    */
+    globjects::Texture * loadRawImage(const std::string & filename) const;
+
+
+protected:
+    std::vector<std::string> m_extensions; /**< List of supported file extensions (e.g., ".bmp") */
```
Comment (scheibel) [inline source/gloperate/include/gloperate/loaders/GlrawTextureLoader.h]: These variables can be static and initialized once.

Comment (scheibel) [inline source/gloperate/include/gloperate/loaders/GlrawTextureLoader.h]: Further, they can be declared and defined as `formatsBySuffix` and `typesBySuffix`, too.

File: source/gloperate/source/loaders/GlrawTextureLoader.cpp [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+    cppassist::DescriptiveRawFile rawFile;
+    if (!rawFile.load(filename))
+        return nullptr;
+
+    const int w = rawFile.intProperty("width");
+    const int h = rawFile.intProperty("height");
+
+    //TODO this should be a unique ptr like this:
+    //auto texture = globjects::Texture::createDefault(gl::GL_TEXTURE_2D);
+    globjects::Texture * texture = new globjects::Texture(gl::GL_TEXTURE_2D);
```
Comment (scheibel) [inline source/gloperate/source/loaders/GlrawTextureLoader.cpp]: Use ```globjects::Texture::createDefault(gl::GL_TEXTURE_2D);``` instead as this call initializes with default sampling parameters.
Using the current interface, a texture is not ready to use in shaders as sampler information are missing.

File: source/gloperate/include/gloperate/loaders/GlrawTextureLoader.h (hunk 21 above)
Comment (chistopher) [inline source/gloperate/include/gloperate/loaders/GlrawTextureLoader.h]: My implementation is consistent with `ShaderLoader`. Should I change that class, too?

Comment (scheibel) [inline source/gloperate/include/gloperate/loaders/GlrawTextureLoader.h]: Sounds good.

**Pull request merged**